File: packger/go/internal/installer/installer.go
package installer

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"runtime"
)

const (
	Version = "v0.1.1"
	Repo    = "pd241008/DevTrace"
)

// EnsureBinary checks if the binary is installed in ~/.devtrace/bin.
// If not, it downloads the correct binary for the current OS/Arch.
func EnsureBinary() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("failed to get home directory: %w", err)
	}

	binDir := filepath.Join(home, ".devtrace", "bin")
	if err := os.MkdirAll(binDir, 0755); err != nil {
		return "", fmt.Errorf("failed to create bin directory: %w", err)
	}

	binName := "devtrace"
	if runtime.GOOS == "windows" {
		binName = "devtrace.exe"
	}

	binPath := filepath.Join(binDir, binName)
	if _, err := os.Stat(binPath); err == nil {
		// Binary already exists
		return binPath, nil
	}

	// Not found, download it
	remoteBinName := getRemoteBinaryName()
	downloadURL := getDownloadURL(remoteBinName)

	fmt.Printf("[DevTrace] Downloading binary for %s/%s from %s...\n", runtime.GOOS, runtime.GOARCH, downloadURL)

	if err := downloadFile(binPath, downloadURL); err != nil {
		return "", fmt.Errorf("failed to download binary: %w\nEnsure you have an internet connection and the release exists.", err)
	}

	// Make executable
	if runtime.GOOS != "windows" {
		if err := os.Chmod(binPath, 0755); err != nil {
			return "", fmt.Errorf("failed to make binary executable: %w", err)
		}
	}

	fmt.Printf("[DevTrace] Successfully installed to %s\n", binPath)
	return binPath, nil
}

func getRemoteBinaryName() string {
	osName := runtime.GOOS
	if osName == "darwin" {
		osName = "macos" // Match our release matrix naming
	}
	
	archName := runtime.GOARCH
	if archName == "amd64" {
		archName = "x64"
	}

	ext := ""
	if runtime.GOOS == "windows" {
		ext = ".exe"
	}

	return fmt.Sprintf("devtrace-%s-%s%s", osName, archName, ext)
}

func getDownloadURL(remoteBinName string) string {
	customURL := os.Getenv("DEVTRACE_BINARY_URL")
	if customURL != "" {
		return fmt.Sprintf("%s/%s", customURL, remoteBinName)
	}
	return fmt.Sprintf("https://github.com/%s/releases/download/%s/%s", Repo, Version, remoteBinName)
}

func downloadFile(filepath string, url string) error {
	out, err := os.Create(filepath)
	if err != nil {
		return err
	}
	defer out.Close()

	resp, err := http.Get(url)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("bad status: %s", resp.Status)
	}

	_, err = io.Copy(out, resp.Body)
	return err
}

File: packger/python/devtrace/installer.py
import os
import platform
import urllib.request
import sys
import stat

VERSION = "v0.1.1"
REPO = "pd241008/DevTrace"

def get_binary_name():
    os_name = platform.system().lower()
    if os_name == "darwin":
        os_name = "macos"
    
    arch_name = platform.machine().lower()
    if arch_name in ["x86_64", "amd64"]:
        arch_name = "x64"
    elif arch_name in ["arm64", "aarch64"]:
        arch_name = "arm64"
        
    ext = ".exe" if os_name == "windows" else ""
    return f"devtrace-{os_name}-{arch_name}{ext}"

def get_download_url(binary_name):
    custom_url = os.environ.get("DEVTRACE_BINARY_URL")
    if custom_url:
        return f"{custom_url}/{binary_name}"
    return f"https://github.com/{REPO}/releases/download/{VERSION}/{binary_name}"

def ensure_binary():
    home = os.path.expanduser("~")
    bin_dir = os.path.join(home, ".devtrace", "bin")
    os.makedirs(bin_dir, exist_ok=True)
    
    binary_name = "devtrace.exe" if platform.system().lower() == "windows" else "devtrace"
    bin_path = os.path.join(bin_dir, binary_name)
    
    if os.path.exists(bin_path):
        return bin_path
        
    remote_binary_name = get_binary_name()
    url = get_download_url(remote_binary_name)
    
    print(f"[DevTrace] Downloading binary for {platform.system()} {platform.machine()} from {url}...")
    
    try:
        with urllib.request.urlopen(url) as response, open(bin_path, 'wb') as out_file:
            data = response.read()
            out_file.write(data)
    except Exception as e:
        print(f"[DevTrace] Failed to download binary: {e}")
        print("[DevTrace] Ensure you have an internet connection and the release exists.")
        sys.exit(1)
        
    if platform.system().lower() != "windows":
        st = os.stat(bin_path)
        os.chmod(bin_path, st.st_mode | stat.S_IEXEC)
        
    print(f"[DevTrace] Successfully installed to {bin_path}")
    return bin_path
